### ISSUE TITLE
Tag completion when editing tiddlers

### ISSUE THREAD

**Issue by prof-web-coding** [Jermolene/TiddlyWiki5#2864]
In the last release the following feature was added: 
- Extended search mechanism to require the search string to be a minimum length, see https://github.com/Jermolene/TiddlyWiki5/commit/989cee50594a76a0f30d1d2124ce9e5e5808c6cf

Would it be possible to add a similar feature for the tag completion when editing tiddlers? I've got more than 2.000 Tiddlers and now it starts to take a couple of seconds until the list of tags is loaded when you want to add a tag to a tiddler.

**Comment by pmario** [comment]
The default search mechanism uses: `$:/config/Search/MinLength` as a config tiddler. ... Do you think we can work with the same value, or should we create a new one?

**Comment by pmario** [comment]
As a workaroiund: 

if you open [$:/core/macros/tag-picker](http://tiddlywiki.com/#%24%3A%2Fcore%2Fmacros%2Ftag-picker) and copy paste the following code into it, you should get what you want. 

```
\define add-tag-actions()
<$action-sendmessage $message="tm-add-tag" $param={{$:/temp/NewTagName}}/>
<$action-deletetiddler $tiddler="$:/temp/NewTagName"/>
\end

\define tag-button()
<$button class="tc-btn-invisible" tag="a">
$(actions)$
<$action-deletetiddler $tiddler="$:/temp/NewTagName"/>
<$macrocall $name="tag-pill" tag=<<tag>>/>
</$button>
\end

\define tag-picker(actions)
<$set name="actions" value="""$actions$""">
<div class="tc-edit-add-tag">
<span class="tc-add-tag-name">
<$keyboard key="ENTER" actions=<<add-tag-actions>>>
<$edit-text tiddler="$:/temp/NewTagName" tag="input" default="" placeholder={{$:/language/EditTemplate/Tags/Add/Placeholder}} focusPopup=<<qualify "$:/state/popup/tags-auto-complete">> class="tc-edit-texteditor tc-popup-handle"/>
</$keyboard>
</span> <$button popup=<<qualify "$:/state/popup/tags-auto-complete">> class="tc-btn-invisible" tooltip={{$:/language/EditTemplate/Tags/Dropdown/Hint}} aria-label={{$:/language/EditTemplate/Tags/Dropdown/Caption}}>{{$:/core/images/down-arrow}}</$button> <span class="tc-add-tag-button">
<$set name="tag" value={{$:/temp/NewTagName}}>
<$button set="$:/temp/NewTagName" setTo="" class="">
$actions$
<$action-deletetiddler $tiddler="$:/temp/NewTagName"/>
{{$:/language/EditTemplate/Tags/Add/Button}}
</$button>
</$set>
</span>
</div>
<div class="tc-block-dropdown-wrapper">
<$reveal state=<<qualify "$:/state/popup/tags-auto-complete">> type="nomatch" text="" default="">
<div class="tc-block-dropdown">
<$list filter="[{$:/temp/NewTagName}minlength{$:/config/Search/MinLength}limit[1]]" emptyMessage="""<div class="tc-search-results">{{$:/language/Search/Search/TooShort}}</div>""" variable="listItem">
<$list filter="[tags[]!is[system]search:title{$:/temp/NewTagName}sort[]]" variable="tag">
<<tag-button>>
</$list></$list>
<hr>
<$list filter="[{$:/temp/NewTagName}minlength{$:/config/Search/MinLength}limit[1]]" emptyMessage="""<div class="tc-search-results">{{$:/language/Search/Search/TooShort}}</div>""" variable="listItem">
<$list filter="[tags[]is[system]search:title{$:/temp/NewTagName}sort[]]" variable="tag">
<<tag-button>>
</$list></$list>
</div>
</$reveal>
</div>
</$set>
\end
```

**Comment by pmario** [comment]
As I wrote. This is a workaround. @Jermolene should I create a PR with the changes. Eventually a new Config value?

**Comment by prof-web-coding** [comment]
Thanks @pmario ! This is exactly what I was looking for and indeed, I would be happy to see this in the 'core' with a new config value. :) Thanks a lot! The issue can be closed. I will just leave it open for the time being so @Jermolene can decide whether he would like to see a PR or not.

**Comment by danielo515** [comment]
I think it should be a New config value. Tags are usually significantly shorter than regular search strings.

**Comment by tobibeer** [comment]
I believe the solution is **not** to use minlength but rather

* to not show the taglist when not even a single character was typed
* to limit the results to, at most 100

**Comment by Jermolene** [comment]
It would be good to make the tag completion popup consistent with the main search behaviour.

Personally, I work a good deal with small wikis where the dropdown is very useful after a single character. So perhaps we can make it a configuration value that defaults to 1?

**Comment by tobibeer** [comment]
As I said, for the tags, I would not force any minlength at all. But, in my wiki there are 1000 tags. So, to have the popup show all tags without a single character being typed... is an issue.

**Comment by pmario** [comment]
>Personally, I work a good deal with small wikis where the dropdown is very useful after a single character. So perhaps we can make it a configuration value that defaults to 1?

I did just test it at tiddlywiki.com ... And even with 1 as a default it feels wrong. I'm used to get a long list. .. Now I don't get anything. That feels wrong. 

I'd like to set the default to 0. ... And users, that need the value can change them. 

There is a "Hidden Settings" tiddler https://tiddlywiki.com/#Hidden%20Settings So we create a new one there.

**Comment by prof-web-coding** [comment]
I think the best option is:

- To introduce a new setting defining the minimum length required, this will make sure that users with really big tiddly wikis (like me) don't have to wait some seconds to get the search results.
- Set the default to 0, so people with small tiddly wikis are not 'disturbed' by this new feature. I believe it is much more common to have a set of small tiddly wikis than big ones like I do

I do not think that the following proposal is the best one

> to limit the results to, at most 100, possibly only 30 (and only sort afterwards)

as I think sometimes you actually are interested in one of the last ones. Just imagine you want to add a Journal tag 2017-09-20 by typing 201-.... you would not see your tiddler until you are at 2017-09 (I know you could just type "7-09" and would probably hit it, but it is just to illustrate the constraint).

An alternative way would be to simply introduce a timer, so that e.g. the results are only displayed if the user has not typed anything for more than one (TBD) second. This would also address all the points above and is actually a behavior I've seen on other systems.

**Comment by tobibeer** [comment]
> I believe it is much more common to have a set of small tiddly wikis than big ones like I do

I wouldn't be guessing. As soon as you have +50 tags, it becomes pointless to show the list without having entered a single character, so I do not at all agree with:

> And even with 1 as a default it feels wrong.

What would possibly be wrong about not showing a pointless list? It's like search showing you all tiddlers before having typed anything.

In fact, I think I have wanted to use an existing tiddler as a tag so often, but adding that surely ain't an option if we kick off results w/o having even typed anything.

**Comment by prof-web-coding** [comment]
@tobibeer So you agree with my proposal? I think it is in line with what you are saying, except the default value of 0 which I think is rather trivial and could be addressed by adding something to the help...

**Comment by pmario** [comment]
I did create a PR: Hidden Setting for Tag Dropdown filter #3037, which sets the default to 0.

The reason for this is, that Software should work with a ["principle of least surprise"](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)

Or users are used to the fact, that clicking into the tag selector shows a list. short or long. ... I did test the new design with tw5.com edtion. And it just feels wrong, if ther is no list at all.

----------

>A typical formulation of the principle, from 1984, is: "If a necessary feature has a high astonishment factor, it may be necessary to redesign the feature."[4]
>
>In general engineering design contexts, the principle can be taken to mean that a component of a system should behave in a manner consistent with how users of that component are likely to expect it to behave; that is, users should not be astonished at the way it behaves.

**Comment by tobibeer** [comment]
The "principle of least surprise" to me means that I don't get a full list of all existing tags blocking my entire user experience only because I want to add a single tag.

Word is not about blank word documents... and neither is TiddlyWiki about some pristine, empty state. What the core should address is the average use case, and that ain't the documentation wiki of some plugin, but some personal notebook / wiki with a big list of titles and tags, eventually. Something you might be working with daily.

That is what we should cater for by default, imho. And that is why for me 0 is not an option, at all. 1 might be an acceptable starting point. Especially since you may want to use prefixed tags, e.g. `#next`. So, only entering `#` would already list you all of those, e.g. `#done`, `#waiting`, etc...

**Comment by danielo515** [comment]
I agree with @pmario here. We should never try to guess how the users will use tiddlywiky, we will fail all the times except one: our own use case.
That said,  I think that changing a well known behavior that we never got a complaint about it in years violates the principle of least surprise. We should favor existing users because TW adoption is not very fast, and I think that keeping things the way it works now is a good way.  There is a proposal to make this configurable, no one is forcing any configuration, so I don't see why it should be changed to a default behavior that nobody will expect.
I usually tag my tiddlers by just opening the tag list and clicking on the desired tags. @tobibeer what works for you will not work for me and what works for us will probably won't work for others. 

Regards

**Comment by tobibeer** [comment]
> We should never try to guess how the users will use tiddlywiky

That's a bit of a curious statement. Based on what do you implement the core then?

**Comment by tobibeer** [comment]
> We should favor existing users because TW adoption is not very fast, and I think that keeping things the way it works now is a good way. 

Also curious, because it was existing users like me who suggested to constrain the (search!) results in the first place. The only complaint about minlength with tags I have is that it's 0 not 1. A minlength bigger than 1 has a few use cases where it's not helpful at all for adding tags.

> I usually tag my tiddlers by just opening the tag list and clicking on the desired tags. @tobibeer what works for you will not work for me and what works for us will probably won't work for others.

It's not about what works for me, but rather about what doesn't. And I highly doubt that you will scroll through 100 tags or even 1000 to find the one you want.

**Comment by saqimtiaz** [comment]
@prof-web-coding @Jermolene  Can be closed. This was resolved in #3190 with the introduction of [Hidden Setting: Tags Minimum Length](https://tiddlywiki.com/#Hidden%20Setting%3A%20Tags%20Minimum%20Length)

**Comment by Jermolene** [comment]
Thanks @saqimtiaz
